amethyst_tiles/src/pass.rs

```diff
@@ -0,0 +1,444 @@
+    ecs::{
+        DispatcherBuilder, Entities, Join, Read, ReadExpect, ReadStorage, System, SystemData, World,
+    },
+    geometry::{Plane, Ray},
+    math::{self, convert, Matrix4, Point2, Point3, Vector2, Vector3, Vector4},
+    transform::Transform,
+    Hidden,
+};
+
+use amethyst_assets::{AssetStorage, Handle};
+use amethyst_rendy::{
+    batch::{GroupIterator, OneLevelBatch, OrderedOneLevelBatch},
+    bundle::{RenderOrder, RenderPlan, RenderPlugin, Target},
+    camera::{ActiveCamera, Camera, Projection},
+    pipeline::{PipelineDescBuilder, PipelinesBuilder},
+    pod::IntoPod,
+    rendy::{
+        command::{QueueId, RenderPassEncoder},
+        factory::Factory,
+        graph::{
+            render::{PrepareResult, RenderGroup, RenderGroupDesc},
+            GraphContext, NodeBuffer, NodeImage,
+        },
+        hal::{
+            self,
+            device::Device,
+            pso::{self, ShaderStageFlags},
+        },
+        mesh::AsVertex,
+        shader::{Shader, ShaderSetBuilder, SpirvShader},
+    },
+    resources::Tint as TintComponent,
+    sprite::{SpriteRender, SpriteSheet},
+    sprite_visibility::SpriteVisibility,
+    submodules::{
+        gather::CameraGatherer, DynamicUniform, DynamicVertexBuffer, FlatEnvironmentSub, TextureId,
+        TextureSub,
+    },
+    types::{Backend, Texture},
+    util,
+};
+use amethyst_window::ScreenDimensions;
+use derivative::Derivative;
+use glsl_layout::AsStd140;
+use std::marker::PhantomData;
+
+use crate::{
+    iters::Region,
+    map::{Map, MapStorage, Tile, TileMap},
+    pod::{TileArgs, TileMapArgs},
+    CoordinateEncoder, MortonEncoder2D,
+};
+
+#[cfg(feature = "profiler")]
+use thread_profiler::profile_scope;
+
+lazy_static::lazy_static! {
+    static ref VERTEX: SpirvShader = SpirvShader::new(
+        include_bytes!("../compiled/tiles.vert.spv").to_vec(),
+        ShaderStageFlags::VERTEX,
+        "main",
+    );
+
+    static ref FRAGMENT: SpirvShader = SpirvShader::new(
+        include_bytes!("../compiled/tiles.frag.spv").to_vec(),
+        ShaderStageFlags::FRAGMENT,
+        "main",
+    );
+
+    static ref SHADERS: ShaderSetBuilder = ShaderSetBuilder::default()
+        .with_vertex(&*VERTEX).unwrap()
+        .with_fragment(&*FRAGMENT).unwrap();
+
+}
+
+/// Trait to describe how rendering tiles may be culled for the tilemap to render
+pub trait DrawTiles2DBounds: 'static + std::fmt::Debug + Send + Sync {
+    /// Returns the region to render the tiles
+    fn bounds<T: Tile, E: CoordinateEncoder>(map: &TileMap<T, E>, world: &World) -> Region;
+}
+
+/// Default bounds that returns the entire tilemap
+#[derive(Default, Debug)]
+pub struct DrawTiles2DBoundsDefault;
+impl DrawTiles2DBounds for DrawTiles2DBoundsDefault {
+    fn bounds<T: Tile, E: CoordinateEncoder>(map: &TileMap<T, E>, world: &World) -> Region {
+        Region::new(
+            Point3::new(0, 0, 0),
+            Point3::from(*map.dimensions() - Vector3::new(1, 1, 1)),
```

Comment:
This method should probably just return a Vector, it is a copy type.

Reply:
I don't underrstand what you mean. I can't represent a 3d bounding box with a Vector?

Reply:
No, i meant that `dimensions` call requires a dereference. But after thinking about it, this is probably preferred as is. So ignore that :)